src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -119,10 +120,21 @@ private IDataEncoder GetDataEncoder(TransferMode requestedTransferMode, Transfer
         }
 
         // TODO: Client return values/tasks?
-        public void On(string methodName, Type[] parameterTypes, Func<object[], Task> handler)
+        public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], object, Task> handler, object state)
         {
-            var invocationHandler = new InvocationHandler(parameterTypes, handler);
-            _handlers.AddOrUpdate(methodName, invocationHandler, (_, __) => invocationHandler);
+            var invocationHandler = new InvocationHandler(parameterTypes, handler, state);
+            if (!_handlers.TryGetValue(methodName, out var callBackList))
```

Comment:
I did it this way so that I would conditionally create the `List<InvocationHandler>`.  If I just call add or update I have to create a new list every time.

Reply:
`AddOrUpdate` has an overload for a factory for the add part, so you don't need to create a new list every time.